CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSVReader.java

```diff
@@ -0,0 +1,114 @@
+package io.vantiq.extsrc.CSVSource;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+
+import io.vantiq.extjsdk.ExtensionWebSocketClient;
+
+public class CSVReader {
+    private static final String MAX_LINES_IN_EVENT = "maxLinesInEvent";
+
+
+    static void sendNotification(String filename, int numPacket , ArrayList<Map<String,String>> file ,   ExtensionWebSocketClient oClient )
+    {
+        Map<String,Object> m = new LinkedHashMap<>();
+        m.put("file", filename);
+        m.put("segment", numPacket);
+        m.put("lines", file.toArray());
+        
+        
+        oClient.sendNotification(m);
+    }
+    static String setFieldName(int i , Map<String,String>schema)
+    {
+        String field = String.format("field%d",i); 
+        if (schema != null)
+        {
+            if (schema.get(field)!= null)
+            {
+                field = schema.get(field) ; 
+            }
+        }
+
+        return field ; 
+    }
+    @SuppressWarnings("unchecked")
+    static public ArrayList<Map<String,String>> execute(String csvFile,  Map<String, Object>     config,   ExtensionWebSocketClient oClient  ) {
+
+        String line = "";
+        int numOfRecords ; 
+        int numOfProcessedLinesInSegment = 0 ; 
+        Map<String,String> schema = null;
+
+        if (config.get("schema") != null)
+        {
+            schema = (Map<String,String>) config.get("schema");
+        }
+
+        String delimiter = ";";
+        if (config.get("delimiter") != null)
+        {
+            delimiter = config.get("delimiter").toString();
+        }
+
+
+        int numLinesInEvent = (int) config.get(MAX_LINES_IN_EVENT);
+
+        ArrayList<Map<String,String>> file = new ArrayList<Map<String,String>>(); 
+
+        try (BufferedReader br = new BufferedReader(new FileReader(csvFile))) {
+
+            numOfRecords = 0 ; 
+            while ((line = br.readLine()) != null) {
+                numOfProcessedLinesInSegment++;
+                // use comma as separator
+                String[] values = line.split(delimiter);
+
+                Map<String, String> lineValues = new HashMap<String, String>(); 
+                
+                for (int i = 0 ; i < values.length ; i++)
+                {
+                    String currField = setFieldName(i,schema);
+                    lineValues.put(currField, values[i]);
+                }
+
+                file.add(lineValues);
+                numOfRecords++ ; 
+
+                if (numOfRecords >= numLinesInEvent)
+                {
+                    //TODO : this is for testing logic. could have been smarter...
+                    if (oClient != null) 
```

Comment:
A better choice here would be to package up the events & return the events.  Then, you should get back a sequence of events & you can check that they're all the right size (as it stands, they won't be if you give a big enough file.). 

As it stands, you only check that it processes the first set correctly.  Which it probably does.

Reply:
if the file is extremely big , it would be better to work on the  number of lines and send event by event immediately , not to accumulate all in memory in order to be sent ,  i didn't limit the number of lines in event as it is subject of the size of the message .

Reply:
OK, but you need some way (and some test) to see that the code handles multiple event cases correctly.  I'm pretty sure that it doesn't.

Reply:
i enhanced the testing for supporting multiple event from a single file. 
look at testToManyEventsForSingleSegment